File: backend/app/schemas/support.py
"""Pydantic v2 schemas for the Customer Support module."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator


class UserRef(BaseModel):
    id: int
    display_name: str


class SLARecordRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    cycle: int
    first_response_due: datetime
    resolution_due: datetime
    first_response_at: datetime | None
    resolved_at: datetime | None
    first_response_breached: bool
    resolution_breached: bool
    is_active: bool


class SLASummary(BaseModel):
    """Compact SLA info embedded in ticket list items."""

    first_response_due: datetime
    resolution_due: datetime
    first_response_breached: bool
    resolution_breached: bool
    warning: bool


class TicketCreate(BaseModel):
    model_config = {"extra": "forbid"}

    subject: str = Field(min_length=1, max_length=255)
    description: str | None = None
    priority: str
    contact_id: int
    assignee_id: int | None = None

    @field_validator("priority")
    @classmethod
    def _validate_priority(cls, v: str) -> str:
        allowed = {"low", "medium", "high", "critical"}
        if v not in allowed:
            raise ValueError(f"priority must be one of {allowed}")
        return v


class TicketUpdate(BaseModel):
    model_config = {"extra": "forbid"}

    status: str | None = None
    priority: str | None = None
    assignee_id: int | None = None
    subject: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None


class AssignRequest(BaseModel):
    assignee_id: int


class TicketRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    ref: str
    subject: str
    description: str | None
    status: str
    priority: str
    contact_id: int | None
    contact_name_snapshot: str
    account_id: int | None
    assignee_id: int | None
    created_by_id: int
    sla: SLARecordRead | None = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_orm_with_ref(cls, ticket: Any, sla: Any | None = None) -> "TicketRead":
        data = {
            "id": ticket.id,
            "ref": f"I2I-CRM-{ticket.seq_number:04d}",
            "subject": ticket.subject,
            "description": ticket.description,
            "status": ticket.status,
            "priority": ticket.priority,
            "contact_id": ticket.contact_id,
            "contact_name_snapshot": ticket.contact_name_snapshot,
            "account_id": ticket.account_id,
            "assignee_id": ticket.assignee_id,
            "created_by_id": ticket.created_by_id,
            "created_at": ticket.created_at,
            "updated_at": ticket.updated_at,
            "sla": SLARecordRead.model_validate(sla) if sla else None,
        }
        return cls(**data)


class TicketSummary(BaseModel):
    """Lightweight item for list responses."""

    id: int
    ref: str
    subject: str
    status: str
    priority: str
    contact_name: str
    assignee_id: int | None
    sla: SLASummary | None
    created_at: datetime
    updated_at: datetime


class TicketListResponse(BaseModel):
    items: list[TicketSummary]
    next_cursor: str | None
    total: int


class ReplyCreate(BaseModel):
    model_config = {"extra": "forbid"}

    body: str = Field(min_length=1, max_length=10_000)
    is_internal: bool = False

    @field_validator("body")
    @classmethod
    def _strip_body(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("body must not be blank")
        return stripped


class ReplyRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    ticket_id: int
    author_id: int
    body: str
    is_internal: bool
    created_at: datetime


class ReplyListResponse(BaseModel):
    items: list[ReplyRead]


class ActivityLogRead(BaseModel):
    model_config = {"from_attributes": True, "populate_by_name": True}

    id: int
    event_type: str
    actor_id: int | None
    event_metadata: dict[str, Any] | None = None
    created_at: datetime


class ActivityLogResponse(BaseModel):
    items: list[ActivityLogRead]


class SLAListResponse(BaseModel):
    items: list[SLARecordRead]

File: frontend/src/types/support.ts
export type TicketStatus = 'open' | 'in_progress' | 'resolved' | 'closed'
export type TicketPriority = 'low' | 'medium' | 'high' | 'critical'

export interface SLARecord {
  id: number
  cycle: number
  first_response_due: string
  resolution_due: string
  first_response_at: string | null
  resolved_at: string | null
  first_response_breached: boolean
  resolution_breached: boolean
  is_active: boolean
}

export interface SLASummary {
  first_response_due: string
  resolution_due: string
  first_response_breached: boolean
  resolution_breached: boolean
  warning: boolean
}

export interface Ticket {
  id: number
  ref: string
  subject: string
  description: string | null
  status: TicketStatus
  priority: TicketPriority
  contact_id: number | null
  contact_name_snapshot: string
  account_id: number | null
  assignee_id: number | null
  created_by_id: number
  sla: SLARecord | null
  created_at: string
  updated_at: string
}

export interface TicketSummary {
  id: number
  ref: string
  subject: string
  status: TicketStatus
  priority: TicketPriority
  contact_name: string
  assignee_id: number | null
  sla: SLASummary | null
  created_at: string
  updated_at: string
}

export interface TicketListResponse {
  items: TicketSummary[]
  next_cursor: string | null
  total: number
}

export interface Reply {
  id: number
  ticket_id: number
  author_id: number
  body: string
  is_internal: boolean
  created_at: string
}

export interface ActivityLogEntry {
  id: number
  event_type: string
  actor_id: number | null
  event_metadata: Record<string, unknown> | null
  created_at: string
}

export interface TicketFilters {
  queue?: 'mine' | 'unassigned' | 'all'
  status?: TicketStatus
  priority?: TicketPriority
  assignee_id?: number
  account_id?: number
  created_after?: string
  created_before?: string
}
